sql/src/test/java/org/apache/druid/sql/calcite/TempDirProducer.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.sql.calcite;
+
+import org.apache.druid.java.util.common.FileUtils;
+
+import java.io.File;
+import java.io.IOException;
+
+/**
+ * Creates a hierarchy of temp dirs.
+ *
+ * All created directories will be accessible until the shutdown of the current JVM.
+ * Installs only a single shutdown hook.
+ */
+public class TempDirProducer
+{
+  private static TempDirProducer INSTANCE = null;
+
+  public static TempDirProducer instance()
+  {
+    if (INSTANCE == null) {
+      INSTANCE = new TempDirProducer();
+    }
+    return INSTANCE;
+  }
+
+  private final File tempDir;
+
+  private TempDirProducer()
+  {
+    tempDir = FileUtils.createTempDir("druid-tempdir");
+    Runtime.getRuntime().addShutdownHook(new Thread()
```

Comment:
Doing the cleanup via a shutdown hook means that, unfortunate usage of this class in a JVM that's running all tests for the entire test suite could end up generating tons and tons and tons of little directory turds that don't really get cleaned up as they go along.

Also, registering an object with a shutdown hook in the constructor is kinda a sad pattern.  It would be much nicer if the test framework could perhaps instantiate one of these and then clean it up as part of its own lifecycle instead of relying on an singleton instance and a shutdown hook for cleanup.

Is there a reason that we cannot create one of these and close it more cleanly than a shutdown hook?

Reply:
forgot to left a comment for this: we've discussed that there is no real need to retain this class as a singleton:
* every Supplier can have its own directory
* adding `Closeable` makes it possible to move the cleanup to `close`
* Suppliers can be closeable too

these changes were done :)